rep-2003.rst

```diff
@@ -0,0 +1,75 @@
+REP: 2003
+Title:  Sensor Data and Map QoS Settings
+Author: Steve Macenski <stevenmacenski at gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 08-Nov-2019
+Post-History: 30-Nov-2022
+
+Abstract
+========
+
+This REP covers the standards surrounding Quality of Service (QoS) settings for common needs in ROS 2.
+This includes sensor drivers and map representations to standardize interfaces of not only message types outlined by REPs such as REP 118, 138, and 145, but also QoS to ensure ROS 2 interoperability.
+
+Specification
+=============
+
+Map Quality of Service
+----------------------
+
+Map providers such as map servers and SLAM implementations are expected to provide all maps over a reliable transient-local topic.
```

Comment:
I think you *need* to discuss map updates here, if nothing else than to draw a box around the [GCR](https://github.com/ros-infrastructure/rep/pull/212#issuecomment-553978699) formats you discuss here. You say "all maps" but that comment seems to point to just some maps. 

I understand that the representation is not part of this discussion, but there are separate message based approaches (namely [`map_msgs::msg::OccupancyGridUpdate`](https://github.com/ros-planning/navigation2/blob/e066ed1dee18daad2c6a95a49e5db4fc6407f77e/nav2_costmap_2d/src/costmap_2d_publisher.cpp#L77-L78)) for handling differently aged subscriptions. The last time I looked at this, the state of the RCL/DDS cluster was such that the [`onNewSubscription` functionality in the TODO here](https://github.com/ros-planning/navigation2/blob/e066ed1dee18daad2c6a95a49e5db4fc6407f77e/nav2_costmap_2d/src/costmap_2d_publisher.cpp#L70) was not (yet?) implemented, although that may have changed. 

The ROS 1 mechanism relied on being able to send a full map message to only new subscriptions, thus ensuring that you could subscribe ONLY to the full message topic and get the most up to date map. By saying that all maps should be transient local, are you also ensuring that all subscribers will always get a full update any time the data changes?

Reply:
> The last time I looked at this, the state of the RCL/DDS cluster was such that the [`onNewSubscription` functionality in the TODO here](https://github.com/ros-planning/navigation2/blob/e066ed1dee18daad2c6a95a49e5db4fc6407f77e/nav2_costmap_2d/src/costmap_2d_publisher.cpp#L70) was not (yet?) implemented, although that may have changed.

As of Iron, the core supports this with the use of "Matched" callbacks.  See http://docs.ros.org/en/iron/Releases/Release-Iron-Irwini.html#matched-events .

Reply:
@DLu costmaps are not maps, so the costmap update topics are out of scope since those are collision avoidance/planning models and not environmental models that this REP is covering. The text is pretty clear in line about the scope that this is covering. But this is a good clarification to make:

- [ ] Explicitly restrict REP to environmental models, not to be confused with costmaps or other map-like structures for collision avoidance or C-space representation transportation

but also the OccGrid / Update topics have a stamped header, so it should not be challenging to disambiguate old updates to be disgarded. I still agree that Transient Local is the right answer even if we are using an update-structure. You wouldn't want (1) the main updated map to come in then start processing and wait on a polling basis for the `update` to come in, then (2) the algorithm could start planning with outdated data. Having it be event based on connection due to the QoS is still a preferable experience. There's certainly some technical decision making about how to handle the case, but this isn't a roadmap for map update schemes, this is a REP on the QoS to communicate them under. 

Unless you disagree that updates should be also transient local, [which we do today for for some time now](https://github.com/ros-planning/navigation2/blob/e066ed1dee18daad2c6a95a49e5db4fc6407f77e/nav2_costmap_2d/src/costmap_2d_publisher.cpp#L68-L79), I think this REP does not require further updates on the topic. Though, some other documents about how to process the updates could be useful to provide a general roadmap for end-users, I don't disagree